core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -160,39 +171,82 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
+    {
+        return executor.submit(() -> {
+            CatalogConnector newCatalog = catalogFactory.createCatalog(catalogProperties);
+            log.debug("Loaded catalog: " + catalogProperties.catalogHandle());
+            return newCatalog;
+        });
+    }
+
     @Override
     public void pruneCatalogs(Set<CatalogHandle> catalogsInUse)
     {
-        List<CatalogConnector> removedCatalogs = new ArrayList<>();
+        Map<CatalogHandle, ListenableFuture<CatalogConnector>> catalogToPrune = new HashMap<>();
         catalogRemovingLock.lock();
         try {
             if (stopped) {
                 return;
             }
-            Iterator<Entry<CatalogHandle, CatalogConnector>> iterator = catalogs.entrySet().iterator();
-            while (iterator.hasNext()) {
-                Entry<CatalogHandle, CatalogConnector> entry = iterator.next();
-                if (!catalogsInUse.contains(entry.getKey())) {
-                    iterator.remove();
-                    removedCatalogs.add(entry.getValue());
+
+            Set<CatalogHandle> catalogsToPrune = Sets.difference(catalogs.keySet(), catalogsInUse);
+            if (catalogsToPrune.isEmpty()) {
+                return;
+            }
+
+            for (CatalogHandle catalogHandleToPrune : catalogsToPrune) {
+                ListenableFuture<CatalogConnector> catalogFuture = catalogs.remove(catalogHandleToPrune);
+
+                if (catalogFuture != null) {
+                    catalogToPrune.put(catalogHandleToPrune, catalogFuture);
                 }
             }
         }
         finally {
             catalogRemovingLock.unlock();
         }
 
-        removedCatalogs.forEach(removedCatalog -> Futures.submit(
-                () -> {
-                    try {
-                        removedCatalog.shutdown();
-                        log.debug("Pruned catalog: %s", removedCatalog.getCatalogHandle());
-                    }
-                    catch (Throwable e) {
-                        log.error(e, "Error shutting down catalog: %s".formatted(removedCatalog));
-                    }
-                },
-                executor).state());
+        pruneCatalogs(catalogToPrune);
+    }
+
+    private void pruneCatalogs(Map<CatalogHandle, ListenableFuture<CatalogConnector>> catalogsToPrune)
+    {
+        CountDownLatch catalogsUnloading = new CountDownLatch(catalogsToPrune.size());
+        for (Map.Entry<CatalogHandle, ListenableFuture<CatalogConnector>> entry : catalogsToPrune.entrySet()) {
+            addCallback(pruneCatalog(entry.getValue()), new FutureCallback<>()
+            {
+                @Override
+                public void onSuccess(CatalogHandle result)
+                {
+                    catalogsUnloading.countDown();
+                }
+
+                @Override
+                public void onFailure(Throwable throwable)
+                {
+                    log.error(throwable, "Could not prune catalog: %s", entry.getKey());
```

Comment:
IIUC, the `throwable` may come from loading failure, or from pruning transformation failure
we should not log loading failure as "Could not prune catalog"

the error handling would be simpler if `addCallback(pruneCatalog(entry.getValue())` is replaced with `addCallback(entry.getValue()` -- ie add listener directly on the loading future and call shutdown in `onSuccess()`